fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/VectorIndexScanOptions.java

```diff
@@ -0,0 +1,226 @@
+        }
+
+        return scanOptionsBuilder.build();
+    }
+
+    @Nonnull
+    public ExplainTokensWithPrecedence explain() {
+        final var explainTokens =
+                new ExplainTokens().addSequence(() -> new ExplainTokens().addCommaAndWhiteSpace(),
+                        optionsMap.entrySet()
+                                .stream()
+                                .map(entry ->
+                                        new ExplainTokens().addIdentifier(entry.getKey().getOptionName()).addKeyword(":")
+                                                .addWhitespace().addToString(entry.getValue()))
+                                .collect(Collectors.toList()));
+
+        return ExplainTokensWithPrecedence.of(explainTokens);
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (o == this) {
+            return true;
+        }
+        if (!(o instanceof VectorIndexScanOptions)) {
+            return false;
+        }
+        final VectorIndexScanOptions that = (VectorIndexScanOptions)o;
+        return Objects.equals(optionsMap, that.optionsMap);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(optionsMap);
+    }
+
+    @Override
+    public String toString() {
+        return explain().getExplainTokens().render(DefaultExplainFormatter.forDebugging()).toString();
+    }
+
+    @Nonnull
+    public static VectorIndexScanOptions.Builder builder() {
+        return new Builder();
+    }
+
+    @Nonnull
+    public static VectorIndexScanOptions empty() {
+        return EMPTY;
+    }
+
+    @Nonnull
+    public static VectorIndexScanOptions fromProto(@Nonnull final PVectorIndexScanOptions vectorIndexScanOptionsProto) {
+        final Map<OptionKey<?>, Object> optionsMap =
+                Maps.newHashMapWithExpectedSize(vectorIndexScanOptionsProto.getOptionEntriesCount());
+        for (int i = 0; i < vectorIndexScanOptionsProto.getOptionEntriesCount(); i ++) {
+            final POptionEntry optionEntryProto = vectorIndexScanOptionsProto.getOptionEntries(i);
+            optionsMap.put(Objects.requireNonNull(optionsNameMap.get(optionEntryProto.getKey())),
+                    LiteralKeyExpression.fromProtoValue(optionEntryProto.getValue()));
+        }
+        return new VectorIndexScanOptions(optionsMap);
+    }
+
+    public static class Builder {
+        @Nonnull
+        private final Map<OptionKey<?>, Object> optionsMap;
+
+        public Builder() {
+            this(ImmutableMap.of());
+        }
+
+        public Builder(@Nonnull final Map<OptionKey<?>, Object> optionsMap) {
+            this.optionsMap = Maps.newHashMap(optionsMap);
+        }
+
+        public <T> Builder putOption(@Nonnull final OptionKey<T> key, T value) {
+            optionsMap.put(key, value);
+            return this;
+        }
+
+        VectorIndexScanOptions build() {
+            return new VectorIndexScanOptions(Maps.newHashMap(optionsMap));
```

Comment:
The comment in the `VectorIndexScanOptions` constructor about using a hash map as it's not `null` averse probably makes a bit more sense here. It also seems like (1) it should use an `ImmutableMap` or `Collections.emptyMap()` if `optionsMap` is empty (or a static `VectorIndexScanOptions`) and (2) it should wrap the result in `Collections.umodifiableMap`